### ISSUE TITLE
Netherfactory loading is broken with compactor and IO refactor 

### ISSUE THREAD

**Issue by ttk2** [Civcraft/FactoryMod#109]
2015-06-25 08:25:10 [INFO] FactoryMod v1.4.0: ERROR: compactorSaves.txt is not a valid file!
2015-06-25 08:25:10 [INFO] FactoryMod v1.4.0: ERROR at line 1 of netherSaves.txt: Factory parse error: null
2015-06-25 08:25:10 [INFO] FactoryMod v1.4.0: Finished initializing FactoryMod Managers.

**Comment by ProgrammerDan** [comment]
Arg, that gives me 0 information to solve. What version of FactoryMod was on Civmain prior to loading the new release?

I can create a NF locally and see what the save file looks like, then start breaking it down to figure out why the IOrefactor broke shit.

Was it the only save file type to fail?

**Comment by ProgrammerDan** [comment]
Ok I think it was running 1.3.5. I'm basically out of time to work on this, but if anyone else does:

Checkout commit #2172f30254b989f99553fd5ef3c469a9619754a0 and build it locally, make a few NFs on a local testing machine, upgrade to 1.4.0 in the pull request, and see what you get.

**Comment by BlackXnt** [comment]
Although the NF doesn't work the furnace is still working and you cant turn it off. Trying to click on the crafting table with a stick will return "No Factory was identified!". I also managed to build a new NF on civtest that works. I assume the problem is that the way the plugin reads the factories from the text file was changed in the new version. I will keep investigating.

**Comment by ProgrammerDan** [comment]
Yes, don't bother with Civtest. 

Bring up a fresh compile of the commit I reference on a local civcraft clone, build two NFs, upgrade to latest pull on rm-patch-2-rb branch, and see what breaks.

**Comment by BlackXnt** [comment]
The text file that saves the nether factories is just gibberish for me in the old and new version.

**Comment by ProgrammerDan** [comment]
Sure. When you create a factory on FM 1.3.5, then upgrade to 1.4.0, do you get the same Factory Parse Error on startup?

**Comment by BlackXnt** [comment]
Yes.

**Comment by BlackXnt** [comment]
I used this build for 1.3.5 : http://civcraft.co:8080/view/Latest/job/FactoryMod-master/107/

And this build for 1.4 : http://civcraft.co:8080/view/Latest/job/FactoryMod-master/

**Comment by ProgrammerDan** [comment]
This http://civcraft.co:8080/view/Latest/job/FactoryMod-master/109/ was the last one up on Civcraft, but no matter, your test was valid.

Alright, it won't be solvable looking at the output file. It'll be necessary to tear into the portions of the code that generate and load the TXT save files, which radically changed between these two releases. 

Good luck. I'll be online after work tonight if unsuccessful.

**Comment by BlackXnt** [comment]
Do you remember in which class the file is loaded and saved?

**Comment by ProgrammerDan** [comment]
Old:  
https://github.com/Civcraft/FactoryMod/blob/2172f30254b989f99553fd5ef3c469a9619754a0/src/com/github/igotyou/FactoryMod/managers/NetherFactoryManager.java

New: https://github.com/ProgrammerDan/FactoryMod/blob/rm-patch-2-rb/src/com/github/igotyou/FactoryMod/persistence/NetherCsvReader.java
and https://github.com/ProgrammerDan/FactoryMod/blob/rm-patch-2-rb/src/com/github/igotyou/FactoryMod/persistence/NetherCsvWriter.java

**Comment by WildWeazel** [comment]
Are you loading this on Civcraft now? I'm not sure we verified old-to-new file loading on Civtest yet.

**Comment by BlackXnt** [comment]
@WildWeazel Yes he is. We forgot to test it...

**Comment by WildWeazel** [comment]
Well the compactor error is expected then, the file does not exist but it should be created at that point. NF error would imply a difference in formatting, but I was just looking at what Dan linked and see no change in the way data is read. The null exception is just perfect...

**Comment by BlackXnt** [comment]
It is hard finding the difference without the ability to read the text file...

**Comment by ProgrammerDan** [comment]
We might just need to inject extra error / debug handling. That specific error that is returned is coming from the read() function, which simply returns a NF object. The source of the error could be literally any line in the read() function, up to and including the NetherFactory constructor.

**Comment by ttk2** [comment]
I will feed it to roruke tonight. Or today if he asks on of the other
admins.

**Comment by BlackXnt** [comment]
@ProgrammerDan I will try looking there / making a build with a lot of debugging msgs.

**Comment by rourke750** [comment]
I'll look at my lunch today.

**Comment by BlackXnt** [comment]
@ProgrammerDan You are probably right. The problem is not just with factories that were build at 1.3.5 but with factories that were build at 1.4 too.

**Comment by erocs** [comment]
Ttk2, if you get to it before me grab the code off the server to decide the
java binary format the factories were previously saved in and forward it
along.

**Comment by ProgrammerDan** [comment]
I think I know the problem, lemme try to point it out, one sec...

**Comment by ProgrammerDan** [comment]
So line 106 in NetherCsvReader: `(NetherFactoryManager) mPlugin.manager.getManager(NetherFactoryManager.class)` is basically asking the plugin for the manager object, which needs to return the NF Manager.

However, the read() function is called by the load() function of the constructor of the same FactoryManager.

E.g. it's asking the plugin for a class instance that isn't done instantiating. This will return null, and cause the exception.

**Comment by ProgrammerDan** [comment]
Quick fix: Change the NFManager requirement in the Constructor to be a FactoryModPlugin. Then refactor inside the NetherFactory to use the `(NetherFactoryManager) mPlugin.manager.getManager(NetherFactoryManager.class)` instead of a local copy of NetherFactoryManager.

I'll clean it up later but I'm 75% sure this is the problem.

**Comment by ProgrammerDan** [comment]
@WildWeazel Looks like I should have tried harder to decouple the NetherFactory from the NetherFactoryManager. Looks to me like the coupling is the problem.

Sequence:

FactoryModPlugin starts. It calls the Constructor of the FactoryManager which will eventually be assigned to paramater "manager" in the plugin.

Inside the FactoryManager constructor, it calls the constructors for each specific factory's manager. The superclass constructor of each specific factory manager constructs a DAO for the correct factory type. Back in the specific factory manager's constructor, it calls the "load" function, actually reading the factories from the file using the DAO.

The DAO for nether factory attempts to get the FactoryModPlugin.manager object and call a function on it. However, at this point in time that field is still null, as the constructor has not returned.

Since the Factory loading now fails, the constructor cleanly ends, assigning the new FactoryManager to the manager field on FactoryModPlugin.

**Comment by WildWeazel** [comment]
Yeah that would do it. The only place the manager reference is used is to find the scaling factor. Instead of that dependency injection it seems cleaner to have the NF manager query the factory just for the basic materials and then calculate scaling itself, since it's the one that knows where the other factories are. Then the factory wouldn't need to know about the manager at all.

**Comment by ProgrammerDan** [comment]
Have the manager keep track of scaling factor? The issue is that the computation occurs as a result of player interaction, so somehow that value needs to get to each factory instance, and be kept fresh.

**Comment by ProgrammerDan** [comment]
Actually, easier fix:

replace line 452 with this:

repairMaterials = repairMaterials.getMultiple(((NetherFactoryManager) FactoryModPlugin.getPlugin().manager.getManager(NetherFactoryManager.class)).getScalingFactor(factoryLocation));

then remove the dependency from constructor.

**Comment by BlackXnt** [comment]
Nice catch. Should I try testing it?

**Comment by ProgrammerDan** [comment]
Yeah if you can make the change and compile locally, go for it. Or make a pull against my branch. There's a bunch of small changes to effectuate the fix, in about 3-4 files, but generally straightforward. If you've got it loaded in eclipse or another IDE all the needed changes will highlight as errors once you alter the constructor for NetherFactory.

**Comment by erocs** [comment]
If you still need to crack open the old file, use
http://civcraft.co/DumpNetherFactories.java.

**Comment by WildWeazel** [comment]
That should work too.

I was thinking getRepairs would be called from the manager anyway, but it seems to only come from ABaseFactory.update so there is no manager in context to do the scaling.

**Comment by BlackXnt** [comment]
@ProgrammerDan I successfully managed to create a NF restart the server and use the NF after the restart. The error doesn't show up anymore. Should I also go back to 1.3.5 make a nether factory and upgrade to 1.4?

**Comment by ProgrammerDan** [comment]
@BlackXnt  Yes!

**Comment by BlackXnt** [comment]
@ProgrammerDan It is working! I can use the factory I built in the 1.3.5 version after I updated to the jar I compiled.

**Comment by ProgrammerDan** [comment]
Sweet. If you can issue a pull request against my rm-patch-2-rb branch, I'll merge it in.

**Comment by ProgrammerDan** [comment]
@BlackXnt Need git assist, or should I expect the pull request shortly?

**Comment by BlackXnt** [comment]
Sorry I had to go for a few mins.

**Comment by ProgrammerDan** [comment]
One small bug then it'll be gold

**Comment by ProgrammerDan** [comment]
Fix merged into pull: https://github.com/ProgrammerDan/FactoryMod/commit/cb35bcb4e0decc4284e759dfe8214337d79e398d

@ttk2 once the build is ready, this should solve the problem. 

Note, the files are fine in their current incarnation. The issue was **not** with reading or writing the save files, the issue was after the save was loaded.

So no need to reset the save files from a backup.

**Comment by WildWeazel** [comment]
@BlackXnt is a hero

**Comment by BlackXnt** [comment]
I literally copy-pasted Dan's solution.

**Comment by ttk2** [comment]
good job everyone, on civtest in a second.

**Comment by ttk2** [comment]
no new config errors to patch tomorrow morning? If so talk about quality control for a 5000+ line addition.

**Comment by ttk2** [comment]
looks like the pull fixes it, copied the prod Nf saves file and started it up, not a peep.

**Comment by ttk2** [comment]
well looks like I was wrong, no stacks anywhere but NF's don't work

**Comment by BlackXnt** [comment]
Did you put the backup file with all the factories?

**Comment by ttk2** [comment]
Dan said I did'nt have too looks like I do

**Comment by BlackXnt** [comment]
ofc you do because when you restarted it cleared the file(i think).

**Comment by ProgrammerDan** [comment]
Good catch @BlackXnt, didn't consider that, that's a terrible practice. I'll have to fix that later, but putting the backups in place will certainly solve it.

**Comment by erocs** [comment]
You need to pull the old nether factory save file out of the backup. Also,
save the current one because some were created today while the map was
empty of other NFs. Those will likely need reimbursement.

**Comment by ProgrammerDan** [comment]
@ttk2 -- actually, check the folder. I think the new code puts the old file as a .bck file in the same folder with the new save. Yesterday's factories are probably still resident in the config folder as a .bck file.

**Comment by ttk2** [comment]
well I got it working, but I didn't save the one there, there was only one line.

**Comment by BlackXnt** [comment]
There is a bug with repairing nether factories. ttk you should be getting stacks.

**Comment by hlprimm** [comment]
Can confirm - New Senntisten nether portal is unable to repair. Diamond blocks in chest, scrolled to "repair," but hitting furnace w/ stick reports "factory is in disrepair." 

It is at 0% and the furnace is not on

**Comment by BlackXnt** [comment]
@ProgrammerDan  

[03:23:25] [Server thread/ERROR]: Could not pass event PlayerInteractEvent to FactoryMod v1.4.0
org.bukkit.event.EventException
	at org.bukkit.plugin.java.JavaPluginLoader$1.execute(JavaPluginLoader.java:305) ~[spigot-1.8.3.jar:git-Spigot-870264a-0a645a2]
	at org.bukkit.plugin.RegisteredListener.callEvent(RegisteredListener.java:62) ~[spigot-1.8.3.jar:git-Spigot-870264a-0a645a2]
	at org.bukkit.plugin.SimplePluginManager.fireEvent(SimplePluginManager.java:502) [spigot-1.8.3.jar:git-Spigot-870264a-0a645a2]
	at org.bukkit.plugin.SimplePluginManager.callEvent(SimplePluginManager.java:487) [spigot-1.8.3.jar:git-Spigot-870264a-0a645a2]
	at org.bukkit.craftbukkit.v1_8_R2.event.CraftEventFactory.callPlayerInteractEvent(CraftEventFactory.java:227) [spigot-1.8.3.jar:git-Spigot-870264a-0a645a2]
	at org.bukkit.craftbukkit.v1_8_R2.event.CraftEventFactory.callPlayerInteractEvent(CraftEventFactory.java:194) [spigot-1.8.3.jar:git-Spigot-870264a-0a645a2]
	at net.minecraft.server.v1_8_R2.PlayerInteractManager.a(PlayerInteractManager.java:107) [spigot-1.8.3.jar:git-Spigot-870264a-0a645a2]
	at net.minecraft.server.v1_8_R2.PlayerConnection.a(PlayerConnection.java:604) [spigot-1.8.3.jar:git-Spigot-870264a-0a645a2]
	at net.minecraft.server.v1_8_R2.PacketPlayInBlockDig.a(SourceFile:40) [spigot-1.8.3.jar:git-Spigot-870264a-0a645a2]
	at net.minecraft.server.v1_8_R2.PacketPlayInBlockDig.a(SourceFile:10) [spigot-1.8.3.jar:git-Spigot-870264a-0a645a2]
	at net.minecraft.server.v1_8_R2.PlayerConnectionUtils$1.run(SourceFile:13) [spigot-1.8.3.jar:git-Spigot-870264a-0a645a2]
	at java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source) [?:1.8.0_25]
	at java.util.concurrent.FutureTask.run(Unknown Source) [?:1.8.0_25]
	at net.minecraft.server.v1_8_R2.SystemUtils.a(SourceFile:60) [spigot-1.8.3.jar:git-Spigot-870264a-0a645a2]
	at net.minecraft.server.v1_8_R2.MinecraftServer.A(MinecraftServer.java:712) [spigot-1.8.3.jar:git-Spigot-870264a-0a645a2]
	at net.minecraft.server.v1_8_R2.DedicatedServer.A(DedicatedServer.java:368) [spigot-1.8.3.jar:git-Spigot-870264a-0a645a2]
	at net.minecraft.server.v1_8_R2.MinecraftServer.z(MinecraftServer.java:651) [spigot-1.8.3.jar:git-Spigot-870264a-0a645a2]
	at net.minecraft.server.v1_8_R2.MinecraftServer.run(MinecraftServer.java:554) [spigot-1.8.3.jar:git-Spigot-870264a-0a645a2]
	at java.lang.Thread.run(Unknown Source) [?:1.8.0_25]
Caused by: java.lang.NullPointerException
	at com.github.igotyou.FactoryMod.Factorys.NetherFactory.getRepairs(NetherFactory.java:448) ~[?:?]
	at com.github.igotyou.FactoryMod.Factorys.ABaseFactory.getAllInputs(ABaseFactory.java:250) ~[?:?]
	at com.github.igotyou.FactoryMod.Factorys.ABaseFactory.checkHasMaterials(ABaseFactory.java:120) ~[?:?]
	at com.github.igotyou.FactoryMod.Factorys.ABaseFactory.togglePower(ABaseFactory.java:167) ~[?:?]
	at com.github.igotyou.FactoryMod.listeners.FactoryModListener.playerInteractionEvent(FactoryModListener.java:241) ~[?:?]
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:1.8.0_25]
	at sun.reflect.NativeMethodAccessorImpl.invoke(Unknown Source) ~[?:1.8.0_25]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source) ~[?:1.8.0_25]
	at java.lang.reflect.Method.invoke(Unknown Source) ~[?:1.8.0_25]
	at org.bukkit.plugin.java.JavaPluginLoader$1.execute(JavaPluginLoader.java:301) ~[spigot-1.8.3.jar:git-Spigot-870264a-0a645a2]

This is the stack I am getting on my local server when trying to repair a nether factory.

**Comment by ProgrammerDan** [comment]
Ok I'll fix it.

**Comment by BlackXnt** [comment]
@ProgrammerDan I got it. Just a small typo.
